consensus/polybft/governance_manager.go

```diff
@@ -0,0 +1,541 @@
+// if governance fork is not enabled, reward distribution is only present at the epoch ending block
+func isRewardDistributionBlock(isFirstBlockOfEpoch, isEndOfEpoch bool,
+	pendingBlockNumber uint64) bool {
+	if forkmanager.GetInstance().IsForkEnabled(chain.Governance, pendingBlockNumber) {
+		return isFirstBlockOfEpoch && pendingBlockNumber > 1
+	}
+
+	return isEndOfEpoch
+}
+
+// getLookbackSizeForRewardDistribution returns lookback size for reward distribution
+// based on if governance fork is enabled or not
+func getLookbackSizeForRewardDistribution(blockNumber uint64) uint64 {
+	if forkmanager.GetInstance().IsForkEnabled(chain.Governance, blockNumber) {
+		return newRewardLookbackSize
+	}
+
+	return oldRewardLookbackSize
+}
+
+// GovernanceManager interface provides functions for handling governance events
+// and updating client configuration based on executed governance proposals
+type GovernanceManager interface {
+	PostBlock(req *PostBlockRequest) error
+	PostEpoch(req *PostEpochRequest) error
+	GetClientConfig() (*PolyBFTConfig, error)
+}
+
+var _ GovernanceManager = (*dummyGovernanceManager)(nil)
+
+// dummyStakeManager is a dummy implementation of GovernanceManager interface
+// used only for unit testing
+type dummyGovernanceManager struct{}
+
+func (d *dummyGovernanceManager) PostBlock(req *PostBlockRequest) error { return nil }
+func (d *dummyGovernanceManager) PostEpoch(req *PostEpochRequest) error { return nil }
+func (d *dummyGovernanceManager) GetClientConfig() (*PolyBFTConfig, error) {
+	return nil, nil
+}
+
+var _ GovernanceManager = (*governanceManager)(nil)
+
+// governanceManager is a struct that saves governance events
+// and updates the client configuration based on executed governance proposals
+type governanceManager struct {
+	logger         hclog.Logger
+	state          *State
+	eventsGetter   *eventsGetter[contractsapi.EventAbi]
+	allForksHashes map[types.Hash]string
+}
+
+// newGovernanceManager is a constructor function for governance manager
+func newGovernanceManager(genesisConfig *PolyBFTConfig,
+	logger hclog.Logger,
+	state *State,
+	blockhain blockchainBackend) (*governanceManager, error) {
+	eventsGetter := &eventsGetter[contractsapi.EventAbi]{
+		blockchain: blockhain,
+		isValidLogFn: func(l *types.Log) bool {
+			return l.Address == genesisConfig.GovernanceConfig.NetworkParamsAddr ||
+				l.Address == genesisConfig.GovernanceConfig.ForkParamsAddr
+		},
+		parseEventFn: parseGovernanceEvent,
+	}
+
+	config, err := state.GovernanceStore.getClientConfig()
+	if config == nil || errors.Is(err, errClientConfigNotFound) {
+		// insert initial config to db if not already inserted
+		if err = state.GovernanceStore.insertClientConfig(genesisConfig); err != nil {
+			return nil, err
+		}
+	} else if err != nil {
+		return nil, err
+	}
+
+	// cache all fork name hashes that we have in code
+	allForkNameHashes := map[types.Hash]string{}
+	stringABIType := abi.MustNewType("tuple(string)")
```

Comment:
Does it make sense to initialize it on the package level rather than every time here?

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/1519/commits/4a92ab07bf4ef242b2a664bf3fa2c71509a7289d